wrangler-proto/src/main/java/io/cdap/wrangler/proto/workspace/v2/Workspace.java

```diff
@@ -38,16 +39,20 @@ public class Workspace {
   // this is for insights page in UI
   private final JsonObject insights;
 
+  private HashMap<String, UserDefinedAction> nullabilityMap;
+
   private Workspace(String workspaceName, String workspaceId, List<String> directives,
                     long createdTimeMillis, long updatedTimeMillis, @Nullable SampleSpec sampleSpec,
-                    JsonObject insights) {
+                    JsonObject insights, HashMap<String, UserDefinedAction> nullabilityMap) {
     this.workspaceName = workspaceName;
     this.workspaceId = workspaceId;
     this.directives = directives;
     this.createdTimeMillis = createdTimeMillis;
     this.updatedTimeMillis = updatedTimeMillis;
     this.sampleSpec = sampleSpec;
     this.insights = insights;
+    this.nullabilityMap = nullabilityMap == null || nullabilityMap.isEmpty() ?
```

Comment:
by convention, only variables annotated as `@Nullable` should ever be null. Instead, should be creating an unmodifiable defensive copy here:

```
this.nullabilityMap = Collections.unmodifiableMap(new HashMap(nullabilityMap));
```

This is to prevent the map from ever changing once the object is created.